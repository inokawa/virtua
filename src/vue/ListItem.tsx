/** @jsxImportSource vue */
import {
  ref,
  defineComponent,
  watch,
  StyleValue,
  PropType,
  VNode,
  NativeElements,
} from "vue";
import { ItemResizeObserver } from "../core/resizer";
import { isRTLDocument } from "../core/environment";

/**
 * @internal
 */
export const ListItem = /*#__PURE__*/ defineComponent({
  props: {
    _children: { type: Object as PropType<VNode>, required: true },
    _resizer: {
      type: Function as PropType<ItemResizeObserver>,
      required: true,
    },
    _index: { type: Number, required: true },
    _offset: { type: Number, required: true },
    _hide: { type: Boolean },
    _isHorizontal: { type: Boolean },
    _isSSR: { type: Boolean },
    _as: { type: String as PropType<keyof NativeElements>, required: true },
  },
  setup(props) {
    const elementRef = ref<HTMLDivElement>();

    // The index may be changed if elements are inserted to or removed from the start of props.children
    watch(
      () => elementRef.value && props._index,
      (_, __, onCleanup) => {
        onCleanup(props._resizer(elementRef.value!, props._index));
      },
      {
        flush: "post",
      }
    );

    return () => {
      const {
        _children: children,
        _offset: offset,
        _hide: hide,
        _isHorizontal: isHorizontal,
        _isSSR: isSSR,
        _as: Element,
      } = props;

      const style: StyleValue = {
        position: hide && isSSR ? undefined : "absolute",
        [isHorizontal ? "height" : "width"]: "100%",
        [isHorizontal ? "top" : "left"]: "0px",
        [isHorizontal ? (isRTLDocument() ? "right" : "left") : "top"]:
          offset + "px",
        visibility: !hide || isSSR ? "visible" : "hidden",
      };
      if (isHorizontal) {
        style.display = "flex";
      }

      return (
        <Element ref={elementRef} style={style}>
          {children}
        </Element>
      );
    };
  },
});
