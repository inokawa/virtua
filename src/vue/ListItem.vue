<script setup lang="ts">
import {
    ref,
    watch,
    NativeElements,
    computed,
    Ref,
} from "vue";
import {
    ItemResizeObserver,
    isRTLDocument,
    StateVersion,
    VirtualStore,
} from "../core/index.js";
import type { ItemProps } from "./utils.js";

const props = defineProps<{
    _stateVersion: number,
    _store: VirtualStore,
    _children: VNode,
    _resizer: ItemResizeObserver,
    _index: number,
    _isHorizontal: boolean,
    _isSSR: boolean;
    _as: keyof NativeElements,
    _itemProps: ReturnType<ItemProps>
}>();

const elementRef = ref<HTMLDivElement>();

const offset = computed(() =>
    props._stateVersion && props._store.$getItemOffset(props._index)
);
const hide = computed(() =>
    props._stateVersion && props._store.$isUnmeasuredItem(props._index)
);

// The index may be changed if elements are inserted to or removed from the start of props.children
watch(
    () => elementRef.value && props._index,
    (_, __, onCleanup) => {
        onCleanup(props._resizer(elementRef.value!, props._index));
    },
    {
        flush: "post",
    }
);
</script>

<template>
    <component :is="props._as" ref="elementRef" :style="{
        contain: 'layout style',
        position: hide && props._isSSR ? undefined : 'absolute',
        [props._isHorizontal ? 'height' : 'width']: '100%',
        [props._isHorizontal ? 'top' : 'left']: '0px',
        [props._isHorizontal ? (isRTLDocument() ? 'right' : 'left') : 'top']: offset + 'px',
        visibility: !hide || props._isSSR ? 'visible' : 'hidden',
        ...(props._itemProps?.style ?? {}),
        display: props._isHorizontal ? 'inline-flex' : undefined,
    }" v-bind="props._itemProps">
        <component :is="props._children" />
    </component>
</template>
