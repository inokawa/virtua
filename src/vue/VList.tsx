/** @jsxImportSource vue */
import {
  defineComponent,
  ComponentOptionsMixin,
  SlotsType,
  ComponentOptionsWithObjectProps,
  ComponentObjectPropsOptions,
  ref,
  VNode,
} from "vue";
import { Virtualizer, VirtualizerHandle } from "./Virtualizer";

interface VListHandle extends VirtualizerHandle {}

const props = {
  /**
   * The data items rendered by this component.
   */
  data: { type: Array, required: true },
  /**
   * Number of items to render above/below the visible bounds of the list. You can increase to avoid showing blank items in fast scrolling.
   * @defaultValue 4
   */
  overscan: { type: Number, default: 4 },
  /**
   * Item size hint for unmeasured items. It will help to reduce scroll jump when items are measured if used properly.
   *
   * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
   * - If set, you can opt out estimation and use the value as initial item size.
   */
  itemSize: Number,
  /**
   * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
   */
  shift: Boolean,
  /**
   * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
   */
  horizontal: Boolean,
  /**
   * A prop for SSR. If set, the specified amount of items will be mounted in the initial rendering regardless of the container size until hydrated.
   */
  ssrCount: Number,
} satisfies ComponentObjectPropsOptions;

export const VList = /*#__PURE__*/ defineComponent({
  props: props,
  emits: ["scroll", "scrollEnd", "rangeChange"],
  setup(props, { emit, expose, slots }) {
    const horizontal = props.horizontal;

    const onScroll = (offset: number) => {
      emit("scroll", offset);
    };
    const onScrollEnd = () => {
      emit("scrollEnd");
    };
    const onRangeChange = (start: number, end: number) => {
      emit("rangeChange", start, end);
    };

    const handle = ref<InstanceType<typeof Virtualizer>>();

    expose({
      get scrollOffset() {
        return handle.value!.scrollOffset;
      },
      get scrollSize() {
        return handle.value!.scrollSize;
      },
      get viewportSize() {
        return handle.value!.viewportSize;
      },
      getItemOffset: (...args) => handle.value!.getItemOffset(...args),
      scrollToIndex: (...args) => handle.value!.scrollToIndex(...args),
      scrollTo: (...args) => handle.value!.scrollTo(...args),
      scrollBy: (...args) => handle.value!.scrollBy(...args),
    } satisfies VListHandle);

    return () => {
      return (
        <div
          style={{
            display: horizontal ? "inline-block" : "block",
            [horizontal ? "overflowX" : "overflowY"]: "auto",
            contain: "strict",
            width: "100%",
            height: "100%",
          }}
        >
          <Virtualizer
            ref={handle}
            data={props.data}
            overscan={props.overscan}
            itemSize={props.itemSize}
            shift={props.shift}
            ssrCount={props.ssrCount}
            horizontal={horizontal}
            onScroll={onScroll}
            onScrollEnd={onScrollEnd}
            onRangeChange={onRangeChange}
          >
            {slots}
          </Virtualizer>
        </div>
      );
    };
  },
} as ComponentOptionsWithObjectProps<
  typeof props,
  VListHandle,
  {},
  {},
  {},
  ComponentOptionsMixin,
  ComponentOptionsMixin,
  {
    /**
     * Callback invoked whenever scroll offset changes.
     * @param offset Current scrollTop or scrollLeft.
     */
    scroll: (offset: number) => void;
    /**
     * Callback invoked when scrolling stops.
     */
    scrollEnd: () => void;
    /**
     * Callback invoked when visible items range changes.
     */
    rangeChange: (
      /**
       * The start index of viewable items.
       */
      startIndex: number,
      /**
       * The end index of viewable items.
       */
      endIndex: number
    ) => void;
  },
  string,
  {},
  string,
  SlotsType<{ default: (arg: { item: any; index: number }) => VNode[] }>
>);
