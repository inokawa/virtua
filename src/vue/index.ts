/**
 * @module vue
 */
export { default as VList } from "./VList.vue";
export type { VListHandle } from "./VList.vue";
export { default as Virtualizer } from "./Virtualizer.vue";
export type { VirtualizerHandle } from "./Virtualizer.vue";
export { default as WindowVirtualizer } from "./WindowVirtualizer.vue";
export type { WindowVirtualizerHandle } from "./WindowVirtualizer.vue";
