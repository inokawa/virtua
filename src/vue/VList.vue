<script setup lang="ts" generic="T">
import { ref, defineEmits, defineProps, defineExpose, VNode } from "vue";
import Virtualizer from "./Virtualizer.vue";
import type { VirtualizerHandle } from "./Virtualizer.vue";
import type { ItemProps } from "./utils.js";

export interface VListHandle extends VirtualizerHandle { }

interface VListProps<T> {
    /**
     * The data items rendered by this component.
     */
    data: T[];
    /**
     * Number of items to render above/below the visible bounds of the list. You can increase to avoid showing blank items in fast scrolling.
     * @defaultValue 4
     */
    overscan?: number;
    /**
     * Item size hint for unmeasured items in pixels. It will help to reduce scroll jump when items are measured if used properly.
     *
     * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
     * - If set, you can opt out estimation and use the value as initial item size.
     */
    itemSize?: number;
    /**
     * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
     */
    shift?: boolean;
    /**
     * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
     */
    horizontal?: boolean;
    /**
     * A prop for SSR. If set, the specified amount of items will be mounted in the initial rendering regardless of the container size until hydrated.
     */
    ssrCount?: number;
    /**
     * A function that provides properties/attributes for item element
     *
     * **This prop will be merged into `item` prop in the future**
     */
    itemProps?: ItemProps;
    /**
     * List of indexes that should be always mounted, even when off screen.
     */
    keepMounted?: readonly number[];
}

defineSlots<{
    default: (props: { item: T, index: number }) => VNode;
}>();
const props = withDefaults(defineProps<VListProps<T>>(), { overscan: 4 });
const emit = defineEmits<{ scroll: [offset: number]; scrollEnd: [] }>();

const horizontal = !!props.horizontal;

const onScroll = (offset: number) => {
    emit("scroll", offset);
};
const onScrollEnd = () => {
    emit("scrollEnd");
};

const handle = ref<VirtualizerHandle | null>(null);

defineExpose<VListHandle>({
    get scrollOffset() {
        return handle.value!.scrollOffset
    },
    get scrollSize() {
        return handle.value!.scrollSize
    },
    get viewportSize() {
        return handle.value!.viewportSize
    },
    findStartIndex: (...args) => handle.value!.findStartIndex(...args),
    findEndIndex: (...args) => handle.value!.findEndIndex(...args),
    getItemOffset: (...args) => handle.value!.getItemOffset(...args),
    getItemSize: (...args) => handle.value!.getItemSize(...args),
    scrollToIndex: (...args) => handle.value!.scrollToIndex(...args),
    scrollTo: (...args) => handle.value!.scrollTo(...args),
    scrollBy: (...args) => handle.value!.scrollBy(...args),
});
</script>

<template>
    <div :style="{
        display: horizontal ? 'inline-block' : 'block',
        [horizontal ? 'overflowX' : 'overflowY']: 'auto',
        contain: 'strict',
        width: '100%',
        height: '100%',
    }">
        <Virtualizer ref="handle" :data="props.data" :overscan="props.overscan" :itemSize="props.itemSize"
            :itemProps="props.itemProps" :shift="props.shift" :ssrCount="props.ssrCount" :horizontal="horizontal"
            :keepMounted="props.keepMounted" @scroll="onScroll" @scrollEnd="onScrollEnd">
            <template v-slot:["default"]="slotProps">
                <slot :name="'default'" v-bind="slotProps" />
            </template>
        </Virtualizer>
    </div>
</template>
