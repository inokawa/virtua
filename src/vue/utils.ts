import { CSSProperties, VNode } from "vue";
import { ItemsRange } from "../core";

/**
 * @internal
 */
export const getKey = (e: VNode, i: number): Exclude<VNode["key"], null> => {
  const key = e.key;
  return key != null ? key : "_" + i;
};

/**
 * @internal
 */
export const isSameRange = (prev: ItemsRange, next: ItemsRange): boolean => {
  return prev[0] === next[0] && prev[1] === next[1];
};

export type ItemProps = (payload: {
  item: any;
  index: number;
}) => { [key: string]: any; style?: CSSProperties; class?: string } | undefined;
