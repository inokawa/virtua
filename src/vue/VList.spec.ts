import { it, expect, describe } from "vitest";
import { defineComponent, h } from "vue";
import { render } from "@testing-library/vue";
import { VList } from "./VList";
import { setupJsDomEnv } from "../../scripts/spec";

const ITEM_HEIGHT = 50;
const ITEM_WIDTH = 100;
const VIEWPORT_HEIGHT = ITEM_HEIGHT * 10;

setupJsDomEnv({
  itemWidth: ITEM_WIDTH,
  itemHeight: ITEM_HEIGHT,
  viewportHeight: VIEWPORT_HEIGHT,
});

const range = (length: number) => Array.from({ length }).map((_, i) => i);

it("should pass attributes to element", () => {
  const wrapper = render(VList, {
    props: {
      data: range(1),
    },
    attrs: {
      id: "id",
      className: "class",
      "tab-index": 0,
      role: "list",
      "aria-label": "test",
      style: { background: "red" },
    },
    slots: { default: (data: any) => h("div", { key: data }, data) },
  });
  expect(wrapper.baseElement).toMatchSnapshot();
});

describe("vertical", () => {
  it("should render 0 children", () => {
    const wrapper = render(VList, {
      props: {
        data: [],
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 1 children", () => {
    const wrapper = render(VList, {
      props: {
        data: range(1),
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 5 children", () => {
    const wrapper = render(VList, {
      props: {
        data: range(5),
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 100 children", () => {
    const wrapper = render(VList, {
      props: {
        data: range(100),
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 1000 children", () => {
    const wrapper = render(VList, {
      props: {
        data: range(1000),
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 10000 children", () => {
    const wrapper = render(VList, {
      props: {
        data: range(10000),
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render component", () => {
    const Comp = defineComponent({
      props: ["data"],
      setup(props) {
        return () => {
          return h("div", props.data);
        };
      },
    });
    const wrapper = render(VList, {
      props: {
        data: range(3),
      },
      slots: { default: (data: any) => h(Comp, { key: data, data }) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render with given width / height", () => {
    const wrapper = render(VList, {
      props: {
        data: range(5),
      },
      attrs: {
        style: { width: "100px", height: "800px" },
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });
});

describe("horizontal", () => {
  it("should render 0 children", () => {
    const wrapper = render(VList, {
      props: {
        data: [],
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 1 children", () => {
    const wrapper = render(VList, {
      props: {
        data: range(1),
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 5 children", () => {
    const wrapper = render(VList, {
      props: {
        data: range(5),
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 100 children", () => {
    const wrapper = render(VList, {
      props: {
        data: range(100),
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 1000 children", () => {
    const wrapper = render(VList, {
      props: {
        data: range(1000),
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 10000 children", () => {
    const wrapper = render(VList, {
      props: {
        data: range(10000),
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render component", () => {
    const Comp = defineComponent({
      props: ["data"],
      setup(props) {
        return () => {
          return h("div", props.data);
        };
      },
    });
    const wrapper = render(VList, {
      props: {
        data: range(3),
        horizontal: true,
      },
      slots: { default: (data: any) => h(Comp, { key: data, data }) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render with given width / height", () => {
    const wrapper = render(VList, {
      props: {
        data: range(5),
        horizontal: true,
      },
      attrs: {
        style: { width: "100px", height: "800px" },
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });
});
