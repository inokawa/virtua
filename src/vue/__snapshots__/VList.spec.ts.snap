// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`horizontal > should render 0 children 1`] = `
<body>
  <div>
    <div
      style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 0px; height: 100%;"
      />
    </div>
  </div>
</body>
`;

exports[`horizontal > should render 1 children 1`] = `
<body>
  <div>
    <div
      style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 40px; height: 100%;"
      >
        <div
          style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
        >
          <div>
            0
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`horizontal > should render 5 children 1`] = `
<body>
  <div>
    <div
      style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 200px; height: 100%;"
      >
        <div
          style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 40px; visibility: hidden; display: flex;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 80px; visibility: hidden; display: flex;"
        >
          <div>
            2
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 120px; visibility: hidden; display: flex;"
        >
          <div>
            3
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 160px; visibility: hidden; display: flex;"
        >
          <div>
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`horizontal > should render 100 children 1`] = `
<body>
  <div>
    <div
      style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 4000px; height: 100%;"
      >
        <div
          style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 40px; visibility: hidden; display: flex;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 80px; visibility: hidden; display: flex;"
        >
          <div>
            2
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 120px; visibility: hidden; display: flex;"
        >
          <div>
            3
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 160px; visibility: hidden; display: flex;"
        >
          <div>
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`horizontal > should render 1000 children 1`] = `
<body>
  <div>
    <div
      style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 40000px; height: 100%;"
      >
        <div
          style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 40px; visibility: hidden; display: flex;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 80px; visibility: hidden; display: flex;"
        >
          <div>
            2
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 120px; visibility: hidden; display: flex;"
        >
          <div>
            3
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 160px; visibility: hidden; display: flex;"
        >
          <div>
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`horizontal > should render 10000 children 1`] = `
<body>
  <div>
    <div
      style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 400000px; height: 100%;"
      >
        <div
          style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 40px; visibility: hidden; display: flex;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 80px; visibility: hidden; display: flex;"
        >
          <div>
            2
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 120px; visibility: hidden; display: flex;"
        >
          <div>
            3
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 160px; visibility: hidden; display: flex;"
        >
          <div>
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`horizontal > should render component 1`] = `
<body>
  <div>
    <div
      style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 120px; height: 100%;"
      >
        <div
          style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 40px; visibility: hidden; display: flex;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 80px; visibility: hidden; display: flex;"
        >
          <div>
            2
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`horizontal > should render with given width / height 1`] = `
<body>
  <div>
    <div
      style="display: inline-block; overflow-x: auto; contain: strict; width: 100px; height: 800px;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 200px; height: 100%;"
      >
        <div
          style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 40px; visibility: hidden; display: flex;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 80px; visibility: hidden; display: flex;"
        >
          <div>
            2
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 120px; visibility: hidden; display: flex;"
        >
          <div>
            3
          </div>
        </div>
        <div
          style="position: absolute; height: 100%; top: 0px; left: 160px; visibility: hidden; display: flex;"
        >
          <div>
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should pass attributes to element 1`] = `
<body>
  <div>
    <div
      aria-label="test"
      class="class"
      id="id"
      role="list"
      style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%; background: red;"
      tab-index="0"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 40px;"
      >
        <div
          style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
        >
          <div>
            0
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render 0 children 1`] = `
<body>
  <div>
    <div
      style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 0px;"
      />
    </div>
  </div>
</body>
`;

exports[`vertical > should render 1 children 1`] = `
<body>
  <div>
    <div
      style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 40px;"
      >
        <div
          style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
        >
          <div>
            0
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render 5 children 1`] = `
<body>
  <div>
    <div
      style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 200px;"
      >
        <div
          style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
        >
          <div>
            2
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 120px; visibility: hidden;"
        >
          <div>
            3
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 160px; visibility: hidden;"
        >
          <div>
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render 100 children 1`] = `
<body>
  <div>
    <div
      style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 4000px;"
      >
        <div
          style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
        >
          <div>
            2
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 120px; visibility: hidden;"
        >
          <div>
            3
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 160px; visibility: hidden;"
        >
          <div>
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render 1000 children 1`] = `
<body>
  <div>
    <div
      style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 40000px;"
      >
        <div
          style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
        >
          <div>
            2
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 120px; visibility: hidden;"
        >
          <div>
            3
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 160px; visibility: hidden;"
        >
          <div>
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render 10000 children 1`] = `
<body>
  <div>
    <div
      style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 400000px;"
      >
        <div
          style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
        >
          <div>
            2
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 120px; visibility: hidden;"
        >
          <div>
            3
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 160px; visibility: hidden;"
        >
          <div>
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render component 1`] = `
<body>
  <div>
    <div
      style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 120px;"
      >
        <div
          style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
        >
          <div>
            2
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render with given width / height 1`] = `
<body>
  <div>
    <div
      style="display: block; overflow-y: auto; contain: strict; width: 100px; height: 800px;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 200px;"
      >
        <div
          style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
        >
          <div>
            2
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 120px; visibility: hidden;"
        >
          <div>
            3
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 160px; visibility: hidden;"
        >
          <div>
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
